File: src/objclint/ClangUtils.m
//
//  ClangUtils.m
//  objclint
//
//  Created by Smirnov on 12/23/12.
//  Copyright (c) 2012 Borsch Lab. All rights reserved.
//

#import "ClangUtils.h"


@implementation ClangUtils


+ (NSString*) filePathForCursor:(CXCursor) cursor {
    CXSourceLocation location = clang_getCursorLocation(cursor);
    
    CXFile file;
    
    clang_getSpellingLocation(location,&file,NULL,NULL,NULL);
    
    CXString fileNameCX = clang_getFileName(file);
    const char* fileNameC = clang_getCString(fileNameCX);
    
    NSString* filePath = nil;
    if(fileNameC)
        filePath = [NSString stringWithUTF8String: fileNameC];
    
    clang_disposeString(fileNameCX);
    
    return filePath;
}

+ (NSString*) tokenKindDescription:(CXTokenKind) tokenKind {
    NSDictionary* tokenKindDescription =
    @{
        @(CXToken_Punctuation) : @"Punctuation",
        @(CXToken_Keyword)     : @"Keyword",
        @(CXToken_Identifier)  : @"Identifier",
        @(CXToken_Literal)     : @"Literal",
        @(CXToken_Comment)     : @"Comment"
    };
    return tokenKindDescription[@(tokenKind)];
}

@end

File: src/objclint/clang-js-bindings/clang-utils.c
//
//  clang-utils.c
//  objclint
//
//  Created by Alexander Smirnov on 1/26/13.
//  Copyright (c) 2013 Alexander Smirnov. All rights reserved.
//
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "clang-utils.h"

const char* getTokenKindSpelling(CXTokenKind tokenKind) {
    switch(tokenKind) {
        case CXToken_Punctuation : return "Punctuation";
        case CXToken_Keyword     : return "Keyword";
        case CXToken_Identifier  : return "Identifier";
        case CXToken_Literal     : return "Literal";
        case CXToken_Comment     : return "Comment";
    }
    return "<unknown token kind>";
}

char* copyCursorFilePath(CXCursor cursor) {
    CXSourceLocation location = clang_getCursorLocation(cursor);
    
    CXFile file;
    
    clang_getSpellingLocation(location,&file,NULL,NULL,NULL);
    
    CXString fileNameCX = clang_getFileName(file);
    const char* fileNameC = clang_getCString(fileNameCX);
    char* result = NULL;
    if(fileNameC){
        unsigned long length = strlen(fileNameC);
        result = (char*)malloc(length+1);
        memset(result, 0, length+1);
        strncpy(result, fileNameC, length);
    }
    
    clang_disposeString(fileNameCX);

    return result;
}